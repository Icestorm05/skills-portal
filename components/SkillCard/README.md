# Skill Card

## Functionality

The skill card displays a card containing data on the skill passed in. It displays the name as the card title. The description is displayed as card text.
Optionally, radio buttons or a readonly list of the skill level and currency can be displayed. Buttons may also be displayed as card actions. These include an `Add`, `Approve`, `Edit` and `Remove` button.

## Dependencies

`<skill-card>` has a dependency on the `<skill-card-status>` and the `<skill-card-radio-buttons>` component.

## Usage

### Props

`<skill-card>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `disabled` | Boolean *optional* | Whether any skill card inputs and buttons are disabled or not. Defaults to `false`.
| `employeeId` | Number *optional* | The employee ID. Required for the add button to work. Defaults to `0`.
| `skill` | Object | The skill object to display.
| `display-add-button` | Boolean *optional* | Displays an add button. Defaults to `false`.
| `display-approve-button` | Boolean *optional* | Displays an approve button. Defaults to `false`.
| `display-edit-button` | Boolean *optional* | Displays an edit button. Defaults to `false`.
| `display-remove-button` | Boolean *optional* | Displays a remove button. Defaults to `false`.
| `display-status-radio-buttons` | Boolean *optional* | Displays skill currency and level radio buttons. Defaults to `false`.
| `display-status-list` | Boolean *optional* | Displays skill currency and level list. Defaults to `false`.

### Schemas

The `skill` prop should match the skill, saved skill or approved skill schema.

#### Skill Schema

| attribute | type | description
| --- | --- | ---
| `name` | String | The name of the skill.
| `description` | String *optional* | The description of the skill.
| `SkillAttributeId` | Number | The unique attribute ID of the skill.

#### Saved Skill Schema

| attribute | type | description
| --- | --- | ---
| `name` | String | The name of the skill.
| `description` | String *optional* | The description of the skill.
| `SkillAttributeId` | Number | The attribute ID of the skill.
| `EmployeeSkillAttributeId` | Number | A link to the SkillAttributeId.
| `EmployeeId` | Number | The ID of the employee who has the skill.
| `EmployeeSkillId` | Number | The unique ID of the saved skill.
| `SkillCurrencyId` | Number | The ID of the currency. 2 maps to `Current`. 3 maps to `Fading`. 4 maps to `Historic`.
| `SkillLevelId` | Number | The ID of the level. 2 maps to `Awareness`. 3 maps to `Practitioner`. 4 maps to `Expert`.

#### Approved Skill Schema

| attribute | type | description
| --- | --- | ---
| `name` | String | The name of the skill.
| `description` | String *optional* | The description of the skill.
| `SkillAttributeId` | Number | The attribute ID of the skill.
| `EmployeeApproveSkillAttributeId` | Number | A link to the SkillAttributeId.
| `EmployeeApprovedSkillId` | Number | The unique ID of the approved skill.
| `EmployeeId` | Number | The ID of the employee who has the skill.
| `SkillCurrencyId` | Number | The ID of the currency. 2 maps to `Current`. 3 maps to `Fading`. 4 maps to `Historic`.
| `SkillLevelId` | Number | The ID of the level. 2 maps to `Awareness`. 3 maps to `Practitioner`. 4 maps to `Expert`.

### Example

```html
<skill-card :skill="skill"
            display-status-radio-buttons
            display-add-button/>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import SkillCard from '~/components/SkillCard/SkillCard.vue';

@Component({
    components: {
        SkillCard,
    },
})
export default class GenericComponent extends Vue {
    private skill = {
        name: 'Test Skill',
        description: 'This is a test skill.',
        SkillAttributeId: 1,
    };
}
```