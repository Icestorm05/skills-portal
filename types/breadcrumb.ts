import { Location } from 'vue-router';

export interface IBreadcrumb {
    link: Location;
    text: string;
}