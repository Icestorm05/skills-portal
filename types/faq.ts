export interface IFaq {
    question: string;
    answers: IFaqAnswer[];
}

export interface IFaqAnswer {
    type: 'p' | 'b' | 'ul' | 'table';
    text: string | string[] | string[][];
}