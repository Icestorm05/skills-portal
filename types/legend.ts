export interface ILegend {
    SkillCategories: ISkillCategoryLegend;
    Levels: ISkillLevelLegend;
    Currencies: ISkillCurrencyLegend;
    fallback: string;
}

export interface ISkillCategoryLegend {
    [key: string]: string;
}

export interface ISkillLevelLegend {
    Awareness: string;
    Practitioner: string;
    Expert: string;
}

export interface ISkillCurrencyLegend {
    Historic: string;
    Fading: string;
    Current: string;
}