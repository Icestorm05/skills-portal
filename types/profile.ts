export interface IProfile {
    Biography: string;
    Company: string;
    Director: string;
    EmailAddress: string;
    EmployeeId: number;
    FirstLineReporting: string;
    FullName: string;
    JobTitle: string;
    KnownAs: string;
    MainStaffNumber: string;
    SecondLineReporting: string;
    Surname: string;
    children: IProfile[];
}