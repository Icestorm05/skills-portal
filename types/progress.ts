export interface IProgress {
    display: boolean;
}