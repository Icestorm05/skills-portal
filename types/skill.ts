import { ECurrency } from 'enums/currency';
import { ELevel } from 'enums/level';

export interface IDescription {
    name: string;
    description: string;
}

export interface ISkillCategory extends IDescription {
    AreaId?: number;
    SubAreaId?: number;
    children: (ISkill|ISkillCategory)[];
}

export interface ISkill extends IDescription {
    SkillAttributeId: number;
    Area: string;

    EmployeeSkillId?: number;
    EmployeeId?: number;
    EmployeeSkillAttributeId?: number;
    SkillCurrencyId?: number;
    SkillLevelId?: number;

    EmployeeApprovedSkillId?: number;
    EmployeeApprovedSkillAttributeId?: number;

    path?: string[];
}
