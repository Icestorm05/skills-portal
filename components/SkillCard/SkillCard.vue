<template lang="pug">
    v-form.SkillCard
        v-card(raised
               height="100%")
            v-layout(column fill-height)
                v-card-title(primary-title).pb-0
                    b.SkillCardName {{ skill.name }}
                v-card-text(v-if="displayPath").text-xs-left.pb-0
                    b(v-if="skill.path.length > 1") Areas: 
                    b(v-else) Area: 
                    span {{ skill.path.join(', ') }}
                v-flex(v-visible="skill.description").pa-0.SkillCardDescriptionContainer
                    v-card-text.pb-0.SkillCardDescription {{ skill.description }}
                v-card-text(:class="{'py-0': displayStatusRadioButtons}")
                    skill-card-radio-buttons(v-if="displayStatusRadioButtons"
                                             :currency="currency"
                                             @update:currency="currencyChanged"
                                             :level="level"
                                             @update:level="levelChanged"
                                             :disabled="disabled").SkillCardRadioButtons
                    skill-card-status(v-if="displayStatusList"
                                      :currency="currency"
                                      :level="level")
                v-card-actions(v-if="displayButton").SkillCardButtons
                        v-btn(v-if="displayAddButton"
                              flat
                              color="primary"
                              :disabled="disabled"
                              aria-label="Add Skill"
                              @click="add").SkillCardAddButton Add
                        v-btn(v-if="displayEditButton"
                              flat
                              color="primary"
                              :disabled="disabled"
                              aria-label="Save Skill"
                              @click="edit").SkillCardEditButton Save
                        v-btn(v-if="displayRemoveButton"
                              flat
                              color="error"
                              :disabled="disabled"
                              aria-label="Remove Skill"
                              @click="remove").SkillCardRemoveButton Remove
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import { Validator } from "vee-validate";
import Vue from "vue";
import { Prop, Provide, Watch } from "vue-property-decorator";

import { ISkill } from "types/skill";

import VBtn from "vuetify/es5/components/VBtn";
import VCard, {
  VCardActions,
  VCardText,
  VCardTitle
} from "vuetify/es5/components/VCard";
import VForm from "vuetify/es5/components/VForm";
import { VFlex, VLayout } from "vuetify/es5/components/VGrid";
import SkillCardRadioButtons from "./SkillCardRadioButtons/SkillCardRadioButtons.vue";
import SkillCardStatus from "./SkillCardStatus/SkillCardStatus.vue";

@Component({
  components: {
    SkillCardRadioButtons,
    SkillCardStatus,
    VBtn,
    VCard,
    VCardActions,
    VCardText,
    VCardTitle,
    VFlex,
    VForm,
    VLayout
  },
  $_veeValidate: {
    validator: "new"
  }
})
export default class SkillCard extends Vue {
  @Provide("validator") public $validator: Validator = this.$validator;

  @Prop({ type: Object, required: true })
  private skill: ISkill;
  @Prop({ type: Number, default: 0 })
  private employeeId: number;
  @Prop({ type: Boolean, default: false })
  private disabled: boolean;

  @Prop({ type: Boolean, default: false })
  private displayAddButton: boolean;
  @Prop({ type: Boolean, default: false })
  private displayEditButton: boolean;
  @Prop({ type: Boolean, default: false })
  private displayRemoveButton: boolean;

  @Prop({ type: Boolean, default: false })
  private displayStatusRadioButtons: boolean;
  @Prop({ type: Boolean, default: false })
  private displayStatusList: boolean;

  @Prop({ type: Boolean, default: false })
  private displayPath: boolean;

  @Prop({ type: Boolean, default: false })
  private autoUpdate: boolean;

  private get displayButton() {
    return (
      this.displayAddButton ||
      this.displayEditButton ||
      this.displayRemoveButton
    );
  }

  private currency = this.skill.SkillCurrencyId
    ? this.skill.SkillCurrencyId
    : null;
  private level = this.skill.SkillLevelId ? this.skill.SkillLevelId : null;

  @Watch("skill")
  private reset(prev, curr) {
    this.currency = this.skill.SkillCurrencyId
      ? this.skill.SkillCurrencyId
      : null;
    this.level = this.skill.SkillLevelId ? this.skill.SkillLevelId : null;
    this.$validator.reset();
  }

  private currencyChanged(currency: number) {
    this.currency = currency;
    this.$emit('currency-change', currency);
    if (this.autoUpdate) {
      this.edit();
    }
  }

  private levelChanged(level: number) {
    this.level = level;
    this.$emit('level-change', level);
    if (this.autoUpdate) {
      this.edit();
    }
  }

  private async add() {
    if (await this.$validator.validateAll()) {
      const skill = {
        ...this.skill,
        EmployeeId: this.employeeId,
        EmployeeSkillAttributeId: this.skill.SkillAttributeId,
        SkillCurrencyId: this.currency,
        SkillLevelId: this.level
      };
      this.$emit("add-skill", skill, this.reset);
    }
  }

  private async edit() {
    if (await this.$validator.validateAll()) {
      const skill = {
        ...this.skill,
        SkillCurrencyId: this.currency,
        SkillLevelId: this.level
      };
      this.$emit("edit-skill", skill, this.reset);
    }
  }

  private async remove() {
    if (await this.$validator.validateAll()) {
      this.$emit("remove-skill", this.skill, this.reset);
    }
  }
}
</script>

<style lang="stylus" scoped>
.SkillCard {
    height: 100%;
}

.SkillCardButtons {
    width: 100%;
    justify-content: center;
}

.SkillCardDescription {
  text-align: left;
}

.SkillCardDescriptionContainer {
    flex-shrink: 0;
}
</style>
