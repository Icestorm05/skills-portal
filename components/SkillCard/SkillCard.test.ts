import SkillCard from "./SkillCard.vue";
import VeeValidate from "vee-validate";
import { createLocalVue, mount } from "@vue/test-utils";
import vVisible from "~plugins/vVisible";

import { ISkill } from "types/skill";

declare const expect: jest.Expect;

const skill: ISkill = {
  SkillAttributeId: 1,
  name: "Test Skill",
  Area: "",
  SkillCurrencyId: 2,
  SkillLevelId: 4,
  description: "This is a test skill."
};

const unsavedSkill: ISkill = {
  SkillAttributeId: 1,
  name: "Test Skill",
  Area: "",
  description: "This is a test skill."
};

const stubs = {
  SkillCardStatus: '<div class="SkillCardStatus"></div>',
  SkillCardRadioButtons: '<div class="SkillCardRadioButtons"></div>'
};

const localVue = createLocalVue();
localVue.use(VeeValidate);
localVue.directive("visible", vVisible);

describe("Skill Card", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayAddButton: true, displayStatusList: true }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("buttons container is not displayed by default", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect((wrapper.vm as any).displayButton).toBeFalsy();
    expect(wrapper.find(".SkillCardButtons").exists()).toBeFalsy();
  });

  test("buttons container is displayed when display add button prop is set to true", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayAddButton: true }
    });
    expect((wrapper.vm as any).displayButton).toBeTruthy();
    expect(wrapper.find(".SkillCardButtons").exists()).toBeTruthy();
  });

  test("buttons container is displayed when display edit button prop is set to true", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayEditButton: true }
    });
    expect((wrapper.vm as any).displayButton).toBeTruthy();
    expect(wrapper.find(".SkillCardButtons").exists()).toBeTruthy();
  });

  test("buttons container is displayed when display remove button prop is set to true", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    expect((wrapper.vm as any).displayButton).toBeTruthy();
    expect(wrapper.find(".SkillCardButtons").exists()).toBeTruthy();
  });

  test("description is visible when their is a description in the skill object", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect(wrapper.find(".SkillCardDescription").isVisible()).toBeTruthy();
  });

  test("description is invisible when their is no description in the skill object", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill: { ...skill, description: "" } }
    });
    expect(wrapper.find(".SkillCardDescription").isVisible()).toBeFalsy();
  });

  test("display add button prop appends an add button", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayAddButton: true }
    });
    expect(wrapper.find(".SkillCardAddButton").exists()).toBeTruthy();
  });

  test("display edit button prop appends an edit button", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayEditButton: true }
    });
    expect(wrapper.find(".SkillCardEditButton").exists()).toBeTruthy();
  });

  test("display remove button prop appends a remove button", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    expect(wrapper.find(".SkillCardRemoveButton").exists()).toBeTruthy();
  });

  test("no skill card status list or radio buttons are displayed by default", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect(wrapper.find(".SkillCardStatus").exists()).toBeFalsy();
    expect(wrapper.find(".SkillCardRadioButtons").exists()).toBeFalsy();
  });

  test("display status list prop appends a skill card status component", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayStatusList: true }
    });
    expect(wrapper.find(".SkillCardStatus").exists()).toBeTruthy();
  });

  test("display status radio buttons prop appends a skill card radio buttons component", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill, displayStatusRadioButtons: true }
    });
    expect(wrapper.find(".SkillCardRadioButtons").exists()).toBeTruthy();
  });

  test("add button click fires add fn", () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayAddButton: true }
    });
    const spy = jest.fn();
    wrapper.setData({ add: spy });
    expect(spy).not.toHaveBeenCalled();
    wrapper.find(".SkillCardAddButton").trigger("click");
    expect(spy).toHaveBeenCalled();
  });

  test("add button click with disabled prop does not emit add-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, disabled: true, displayAddButton: true }
    });
    wrapper.find(".SkillCardAddButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("add-skill")).toBeFalsy();
  });

  test("add button click with failing validation rules does not emit add-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayAddButton: true }
    });
    wrapper.vm.$validator.attach({ name: "test", rules: "required" });
    await wrapper.vm.$validator.validate("test", "");
    wrapper.find(".SkillCardAddButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("add-skill")).toBeFalsy();
  });

  test("add button click emits add-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayAddButton: true }
    });
    wrapper.find(".SkillCardAddButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("add-skill")).toBeTruthy();
  });

  test("add button click emits skill with default attributes added", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill, displayAddButton: true }
    });
    wrapper.find(".SkillCardAddButton").trigger("click");
    await wrapper.vm.$nextTick();
    const savedSkill: ISkill = wrapper.emitted("add-skill")[0][0] as any;
    expect(savedSkill.EmployeeId).toEqual(0);
    expect(savedSkill.EmployeeSkillAttributeId).toEqual(
      unsavedSkill.SkillAttributeId
    );
    expect(savedSkill.SkillCurrencyId).toEqual(null);
    expect(savedSkill.SkillLevelId).toEqual(null);
  });

  test("add button click emits skill with employee id prop added", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { employeeId: 1, skill: unsavedSkill, displayAddButton: true }
    });
    wrapper.find(".SkillCardAddButton").trigger("click");
    await wrapper.vm.$nextTick();
    const savedSkill: ISkill = wrapper.emitted("add-skill")[0][0] as any;
    expect(savedSkill.EmployeeId).toEqual(1);
  });

  test("add button click emits skill with set currency and level added", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill, displayAddButton: true }
    });
    wrapper.find(".SkillCardAddButton").trigger("click");
    wrapper.setData({ currency: 2, level: 4 });
    await wrapper.vm.$nextTick();
    const savedSkill: ISkill = wrapper.emitted("add-skill")[0][0] as any;
    expect(savedSkill.SkillCurrencyId).toEqual(2);
    expect(savedSkill.SkillLevelId).toEqual(4);
  });

  test("edit button click fires add fn", () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayEditButton: true }
    });
    const spy = jest.fn();
    wrapper.setData({ edit: spy });
    expect(spy).not.toHaveBeenCalled();
    wrapper.find(".SkillCardEditButton").trigger("click");
    expect(spy).toHaveBeenCalled();
  });

  test("edit button click with disabled prop does not emit edit-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, disabled: true, displayEditButton: true }
    });
    wrapper.find(".SkillCardEditButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("edit-skill")).toBeFalsy();
  });

  test("edit button click with failing validation rules does not emit edit-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayEditButton: true }
    });
    wrapper.vm.$validator.attach({ name: "test", rules: "required" });
    await wrapper.vm.$validator.validate("test", "");
    wrapper.find(".SkillCardEditButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("edit-skill")).toBeFalsy();
  });

  test("edit button click emits edit-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayEditButton: true }
    });
    wrapper.find(".SkillCardEditButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("edit-skill")).toBeTruthy();
  });

  test("edit button click emits skill with default currency and level added", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill, displayEditButton: true }
    });
    wrapper.find(".SkillCardEditButton").trigger("click");
    await wrapper.vm.$nextTick();
    const editedSkill: ISkill = wrapper.emitted("edit-skill")[0][0] as any;
    expect(editedSkill.SkillCurrencyId).toBeNull();
    expect(editedSkill.SkillLevelId).toBeNull();
  });

  test("edit button click emits skill with amended currency and level added", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill, displayEditButton: true }
    });
    wrapper.find(".SkillCardEditButton").trigger("click");
    wrapper.setData({ currency: 2, level: 4 });
    await wrapper.vm.$nextTick();
    const editedSkill: ISkill = wrapper.emitted("edit-skill")[0][0] as any;
    expect(editedSkill.SkillCurrencyId).toEqual(2);
    expect(editedSkill.SkillLevelId).toEqual(4);
  });

  test("remove button click fires add fn", () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    const spy = jest.fn();
    wrapper.setData({ remove: spy });
    expect(spy).not.toHaveBeenCalled();
    wrapper.find(".SkillCardRemoveButton").trigger("click");
    expect(spy).toHaveBeenCalled();
  });

  test("remove button click with disabled prop does not emit remove-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, disabled: true, displayRemoveButton: true }
    });
    wrapper.find(".SkillCardRemoveButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("remove-skill")).toBeFalsy();
  });

  test("remove button click with failing validation rules does not emit remove-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    wrapper.vm.$validator.attach({ name: "test", rules: "required" });
    await wrapper.vm.$validator.validate("test", "");
    wrapper.find(".SkillCardRemoveButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("remove-skill")).toBeFalsy();
  });

  test("remove button click emits remove-skill event", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    wrapper.find(".SkillCardRemoveButton").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("remove-skill")).toBeTruthy();
  });

  test("remove button click emits skill", async () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill, displayRemoveButton: true }
    });
    wrapper.find(".SkillCardRemoveButton").trigger("click");
    await wrapper.vm.$nextTick();
    const skillToRemove: ISkill = wrapper.emitted("remove-skill")[0][0] as any;
    expect(skillToRemove.name).toEqual(skill.name);
  });

  test("currency data binds to skill currency id", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect(wrapper.vm.$data.currency).toEqual(skill.SkillCurrencyId);
  });

  test("currency data equals null with no skill currency id", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill: unsavedSkill }
    });
    expect(wrapper.vm.$data.currency).toBeNull();
  });

  test("level data binds to skill level id", () => {
    const wrapper = mount(SkillCard, { stubs, propsData: { skill } });
    expect(wrapper.vm.$data.level).toEqual(skill.SkillLevelId);
  });

  test("level data equals null with no skill level id", () => {
    const wrapper = mount(SkillCard, {
      stubs,
      propsData: { skill: unsavedSkill }
    });
    expect(wrapper.vm.$data.level).toBeNull();
  });

  test("reset function sets currency back to null if no skill currency id", () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill }
    });
    expect(wrapper.vm.$data.currency).toBeNull();
    wrapper.setData({ currency: 2 });
    expect(wrapper.vm.$data.currency).toEqual(2);
    (wrapper.vm as any).reset();
    expect(wrapper.vm.$data.currency).toBeNull();
  });

  test("reset function sets currency back to skill currency id", () => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { skill } });
    expect(wrapper.vm.$data.currency).toEqual(skill.SkillCurrencyId);
    wrapper.setData({ currency: 3 });
    expect(wrapper.vm.$data.currency).toEqual(3);
    (wrapper.vm as any).reset();
    expect(wrapper.vm.$data.currency).toEqual(skill.SkillCurrencyId);
  });

  test("reset function sets level back to null if no skill level id", () => {
    const wrapper = mount(SkillCard, {
      localVue,
      stubs,
      propsData: { skill: unsavedSkill }
    });
    expect(wrapper.vm.$data.level).toBeNull();
    wrapper.setData({ level: 4 });
    expect(wrapper.vm.$data.level).toEqual(4);
    (wrapper.vm as any).reset();
    expect(wrapper.vm.$data.level).toBeNull();
  });

  test("reset function sets level back to skill level id", () => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { skill } });
    expect(wrapper.vm.$data.level).toEqual(skill.SkillLevelId);
    wrapper.setData({ level: 3 });
    expect(wrapper.vm.$data.level).toEqual(3);
    (wrapper.vm as any).reset();
    expect(wrapper.vm.$data.level).toEqual(skill.SkillLevelId);
  });

  test("updating skills prop resets the currency and level data values", () => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { skill } });
    wrapper.setData({ currency: 2, level: 4 });
    expect(wrapper.vm.$data.currency).toEqual(2);
    expect(wrapper.vm.$data.level).toEqual(4);
    wrapper.setProps({ skill: unsavedSkill });
    expect(wrapper.vm.$data.currency).toBeNull();
    expect(wrapper.vm.$data.level).toBeNull();
  });

  test('updating currency emits currency-change event', () => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { skill } });
    expect(wrapper.emitted('currency-change')).toBeFalsy();
    (wrapper.vm as any).currencyChanged(3);
    expect(wrapper.emitted('currency-change')).toEqual([[3]]);
  });

  test('updating currency with auto update prop emits edit-skill event', async() => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { autoUpdate: true, skill } });
    expect(wrapper.emitted('edit-skill')).toBeFalsy();
    (wrapper.vm as any).currencyChanged(3);
    await wrapper.vm.$nextTick();
    const updatedSkill: ISkill = wrapper.emitted('edit-skill')[0][0] as any;
    expect(updatedSkill.SkillCurrencyId).toEqual(3);
  });

  test('updating level emits level-change event', () => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { skill } });
    expect(wrapper.emitted('level-change')).toBeFalsy();
    (wrapper.vm as any).levelChanged(3);
    expect(wrapper.emitted('level-change')).toEqual([[3]]);
  });

  test('updating level with auto update prop emits edit-skill event', async() => {
    const wrapper = mount(SkillCard, { localVue, stubs, propsData: { autoUpdate: true, skill } });
    expect(wrapper.emitted('edit-skill')).toBeFalsy();
    (wrapper.vm as any).levelChanged(3);
    await wrapper.vm.$nextTick();
    const updatedSkill: ISkill = wrapper.emitted('edit-skill')[0][0] as any;
    expect(updatedSkill.SkillLevelId).toEqual(3);
  });
});
