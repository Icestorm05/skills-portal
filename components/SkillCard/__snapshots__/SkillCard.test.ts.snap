// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill Card snapshot has not changed 1`] = `
<form novalidate="novalidate" class="SkillCard">
    <div class="card card--raised" style="height: 100%;">
        <div class="layout column fill-height">
            <div class="card__title pb-0 card__title--primary">
                <b class="SkillCardName">Test Skill</b>
            </div>
            <!---->
            <div class="flex pa-0 SkillCardDescriptionContainer" style="visibility: visible;">
                <div class="card__text pb-0 SkillCardDescription">This is a test skill.</div>
            </div>
            <div class="card__text">
                <!---->
                <div class="SkillCardStatus" currency="2" level="4"></div>
            </div>
            <div class="card__actions SkillCardButtons">
                <button type="button" class="SkillCardAddButton btn btn--flat primary--text" aria-label="Add Skill">
                    <div class="btn__content">Add</div>
                </button>
                <!---->
                <!---->
            </div>
        </div>
    </div>
</form>
`;
